pkg/config/config.go

```diff
@@ -130,21 +132,48 @@ func (c *ConfigInfo) writeToFile() error {
 }
 
 // SetConfiguration modifies Odo configurations in the config file
-// as of now only being used for updatenotification
-func (c *ConfigInfo) SetConfiguration(parameter string, value bool) error {
+// as of now being used for timeout, updatenotification
+func (c *ConfigInfo) SetConfiguration(parameter string, value interface{}) error {
 	switch parameter {
+
+	case "timeout":
+		typedval, ok := value.(int)
+		if typedval == 0 {
+			fmt.Println("Cannot set timeout to 0 setting it to the minimum value(1sec)")
+			typedval = 1
+		}
+		if ok != true {
+			return errors.Errorf("type assertion error: unable to set %s", parameter)
+		}
+		c.OdoSettings.Timeout = typedval
+
 	case "updatenotification":
-		c.OdoSettings.UpdateNotification = &value
+		typedval, ok := value.(bool)
+		if ok != true {
+			return errors.Errorf("type assertion error: unable to set %s", parameter)
+		}
+		c.OdoSettings.UpdateNotification = &typedval
+
 	default:
 		return errors.Errorf("unknown parameter :'%s' is not a parameter in odo config", parameter)
 	}
+
 	err := c.writeToFile()
 	if err != nil {
 		return errors.Wrapf(err, "unable to set %s", parameter)
 	}
 	return nil
 }
 
+// GetTimeout returns the value of Timeout from config
+func (c *ConfigInfo) GetTimeout() int {
+	// minimum timeout value is 1
+	if c.OdoSettings.Timeout == 0 {
+		return 1
```

Comment:
Just a curious question, why don’t we set these with defaults in config file ? Is it to avoid the risk of user accidentally deleting it (and then odo fails?)

Reply:
I prefer default value coming from code instead of config file, hence this

Reply:
hmm, both have their own advantages and disadvantages.. but I tend to bend in favor of having it in config file probably because of the following reasons:
* it does not involve any changes to the code if the defaults were to ever change or even their values to non-default ones intentionally. Because code changes involve that whole build-release-update cycle
* exposes to the user the list of possible configurables by way of just viewing the config file or even a command that can simply read the config file and dump it to the cmd o/p
* reduces hard coding in the code which is in my opinion very important

but comes with some side-effects ;) like:
* what if user accidentally deletes the config file
* what if user screws up the config file if he manually edits it

Now the above side-effects can be worked around with some overhead(one such way could be don't allow the user to modify the config file directly by creating odo user as part of odo installation give ownership of config file to only odo user but I agree that only makes our current installation complex. But its easily achievable by way of writing an rpm spec and generating rpm out of it(which we don't do) so this brings me to the next question -- why not?) Expose a command to also list config entries in config file.

Reply:
in both ways we have to hardcode the minimum value, in code

Reply:
how/why in code ?

Reply:
> how/why in code ?

see even if we look into the linux code, the defults are either constants or Macros, 

> 
> Because default config would be in code even if it is a separate file? But think about it this way, the changes to defaults will be localised to one file + any such change will not necessarily warrant an urgent release.. because it can always be worked around by the user by just a change in his version of the config file

I dont know how are we going to alter an already existing default value in a config file,
in fact we can change it easily if the defaults are coming from code

Reply:
> > how/why in code ?
> 
> see even if we look into the linux code, the defults are either constants or Macros,
> 

hmm ...

> > Because default config would be in code even if it is a separate file? But think about it this way, the changes to defaults will be localised to one file + any such change will not necessarily warrant an urgent release.. because it can always be worked around by the user by just a change in his version of the config file
> 
> I dont know how are we going to alter an already existing default value in a config file,
> in fact we can change it easily if the defaults are coming from code

Just like you edit any other value in the config file :) How is it any different? (It is only that the config file, in case of my comment being incoporated, will always have an entry immaterial of whether the user chooses to override it or not... and the code is free from any hard-coded default values or atleast concentrated in an expected single source file..) The code just picks the values from the config file without making any assumptions whether the value is default or user chosen ...

Or at the least I would be atleast happy if the defaults we maintain currently alongside the code, be atleast moved to a separate file which would be intended to serve as the single point of updates to any configurable default values and/or constants... without which, as the code matures, its going to be a challenging task, if we maintain the constants in many places in the code...

Feel free to ignore the comment if it is the only one blocking the PR as this is probably not even the intention of the PR... and we can take it up whenever appropriate in the future... I am ok with it...

Reply:
It would probably make sense to have this as const https://github.com/redhat-developer/odo/blob/d48d812a6bae054b27cf375ab772a0f1801cb1a7/pkg/config/config.go#L14

Would that solve it @anmolbabu ? I'm not sure what config file you mean :-(

Reply:
@kadel I meant the config file as in $HOME/.kube/odo..

So, timeout is anyway a configurable field and all I meant is just moving any configurable value out of code and into a default config file that is packaged with code but is simply copied to the required config file location may be as part of config init in our code.. That way, the default values for configurables are always isolated to a separate file and restoring them is also not a difficult task... And whenever listing of tunables is required we simply read the config file and put it forth(as the config file would then contain all tunables with wither defaults or user tuned values) and there are not 2 separate lists maintained its just  1 single tunable list... Currently I see that, a config value doesn't appear in config file until user has not tuned it and the defaults come from code regions that can be spread around...

It would atleast be nice to isolate it(default values of tunables and any constants related to them) to constants in separate file.. But, I am ok with anything I just wanted to understand why it can't be isolated to separate file and any configurable value tuned/default appears in that separate file and can be located easily in a single place that affects/alters everywhere else that the constant is/needs to be used...

Reply:
> @kadel I meant the config file as in $HOME/.kube/odo..

what if the user is running odo for the first time and there is no config, or if the value is not specified in the config? What value should be used then? Some default value will have to be in the code.

Reply:
@kadel The config file with default values(as I propose now) will be copied to the config file location at the very first odo command invocation(or exactly where the config file is created currently it would be)... The only difference is that the separate file with all tunables with their default values, maintained as part of code will be now simply moved/copied to the config directory...

Reply:
> > @kadel I meant the config file as in $HOME/.kube/odo..
> 
> what if the user is running odo for the first time and there is no config, or if the value is not specified in the config? What value should be used then? Some default value will have to be in the code.

@kadel, I agree even with my current proposal the config is maintained as part of code base only but my point is only its isolation into a separate file(only to avoid the defaults spreading to more than 1 regions of code) and such that any tunable is always present in config file...

Reply:
@anmolbabu so what do you propose?
Moving this to something like a separate package that will just contain constants like https://github.com/redhat-developer/odo/blob/d48d812a6bae054b27cf375ab772a0f1801cb1a7/pkg/config/config.go#L15 and import it here?

It will make those values public, even if they don't have to be :-(

Reply:
> @anmolbabu so what do you propose?
> Moving this to something like a separate package that will just contain constants like
> 
> [odo/pkg/config/config.go](https://github.com/redhat-developer/odo/blob/d48d812a6bae054b27cf375ab772a0f1801cb1a7/pkg/config/config.go#L15)
> 
> Line 15 in [d48d812](/redhat-developer/odo/commit/d48d812a6bae054b27cf375ab772a0f1801cb1a7)
> 
>  configEnvName  = "ODOCONFIG" 
> and import it here?
> It will make those values public, even if they don't have to be :-(

@kadel My proposal is to create a file called odo/odo.conf(name only for indication can be anything more appropriate) and have the default config in it with all tunables and with their respective default values... As part of https://github.com/redhat-developer/odo/blob/master/pkg/config/config.go#L75 copy odo/odo.conf to config path($HOME/.kube/odo) if it does not currently exist... and from any region of code apart from odo/odo.conf remove any default value for any config.. 

But I agree its something not really the aim of this PR(and I really don't want to block this PR for this). So, in my opinion, we can track this as part of a different issue based on its priority but all I wanted is, to just bring this up and learn of any reason/s that didn't allow us to do this... You see I love Odo and don't want to stay away from any info/perspective that takes me even an inch closer to Odo... :)

Reply:
It makes sense @anmolbabu, can you please open a new issue and describe your proposal there?

Reply:
Ack @kadel

Reply:
From my conversation with @kadel it seems distributing binary is the most common(and convenient) way of distributing go packages and packaging conf text files with binary seems not cleanly supported by golang making my above comments practically not viable at the moment. So, until we find a better way, we probably will need to live with the way it is currently...